File: problems/037-dependency-tree-knapsack/solutions/python/main.py
import sys

t = list(map(int, sys.stdin.buffer.read().split()))
it = iter(t)
N, C = next(it), next(it)
ch = [[] for _ in range(N + 1)]
size = [0] * (N + 1)
value = [0] * (N + 1)
for i in range(1, N + 1):
    p, size[i], value[i] = next(it), next(it), next(it)
    ch[p].append(i)
order = []
after = []


def dfs(u):
    pos = len(order)
    order.append(u)
    after.append(0)
    for v in ch[u]:
        dfs(v)
    after[pos] = len(order)


for u in ch[0]:
    dfs(u)
dp = [[0] * (C + 1) for _ in range(N + 1)]
for i in range(N - 1, -1, -1):
    u = order[i]
    skip = dp[after[i]]
    take = dp[i + 1]
    row = dp[i]
    w = size[u]
    v = value[u]
    for c in range(C + 1):
        row[c] = max(skip[c], v + take[c - w] if c >= w else 0)
print(dp[0][C])

File: problems/037-dependency-tree-knapsack/solutions/rust/main.rs
use std::io::{self, Read};
fn dfs(u: usize, ch: &Vec<Vec<usize>>, order: &mut Vec<usize>, after: &mut Vec<usize>) {
    let pos = order.len();
    order.push(u);
    after.push(0);
    for &v in &ch[u] {
        dfs(v, ch, order, after);
    }
    after[pos] = order.len();
}
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut t = s.split_whitespace();
    let n: usize = t.next().unwrap().parse().unwrap();
    let cmax: usize = t.next().unwrap().parse().unwrap();
    let mut ch = vec![Vec::new(); n + 1];
    let mut size = vec![0usize; n + 1];
    let mut value = vec![0u64; n + 1];
    for i in 1..=n {
        let p: usize = t.next().unwrap().parse().unwrap();
        size[i] = t.next().unwrap().parse().unwrap();
        value[i] = t.next().unwrap().parse().unwrap();
        ch[p].push(i);
    }
    let (mut order, mut after) = (Vec::new(), Vec::new());
    for &u in &ch[0] {
        dfs(u, &ch, &mut order, &mut after);
    }
    let mut dp = vec![vec![0u64; cmax + 1]; n + 1];
    for i in (0..n).rev() {
        let u = order[i];
        for c in 0..=cmax {
            dp[i][c] = dp[after[i]][c];
            if c >= size[u] {
                dp[i][c] = dp[i][c].max(value[u] + dp[i + 1][c - size[u]]);
            }
        }
    }
    println!("{}", dp[0][cmax]);
}
